cataclysmbnteam/Cataclysm-BN
# feat(UI): [CR]Per-character default context keybindings for mutations

#### Summary

Implement keybindings specific to a character, working in default context, but bound from a more specific one.

#### Purpose of change

Being able to bind abilities, such as mutations, to default context, would help a lot with QoL of certain "moves".
For example, using a gun vs using finger laser bionic.

Since there are rather few "free" keybinds and mutations and bionics are supposed to vary between characters, those keybinds should be per-character, not global to the game.

Fixes #1059

#### Describe the solution

Per character keybindings are a separate set of keybindings applied on top of existing local+global ones.
They are saved and loaded with a character and overwrite regular local+global binds.
Character keybindings also exist in local and global variants.

Mutation UI, and for now only mutation UI, generates actions (the "thing you bind to a key") for individual mutations. Unlike static ones, those are world-dependent, similar to item actions.
Unlike item actions, those actions have to be inserted into input manager, since they are "seen" outside their parent context.
When any of those actions is bound in mutation UI, it will also appear in default mode - the keybind set used for character actions when there are no menus open and keypad moves the character etc. If a mutation action is executed, it works as if it was activated from mutation UI, but doesn't require opening mutation UI.

Shortcomings:
* [ ] Generated mutation actions should get a struct. Currently they are stringly typed, created by string concatenation and parsed from a string, which is obviously bad practice. This is to minimize conflict with Android builds, which have their own input management code. I'll fix it later, when I get it working and get my Android environment good enough to test things in it
* [ ] Need an idea how to handle world-dependent global (non-character) keybinds in a world that lacks proper dependencies. Error? Skip? Refuse to bind? Allow binding, but warn? Silently load and just have an error string for those cases?
* [ ] Character keybinds on top of non-character ones should start working...

#### Describe alternatives you've considered

For ranged mutations/bionics, a hardcoded "configure character to default this thing when firing" button would be useful.
Could also cover throwing.

I could have picked a different UI to work with here. For example, bionics or item actions.

#### Testing

Ideally, all of this should work:
* Debug in butterfly wings and vomit3 mutations
* Open mutation UI, then keybindings
* Add a character global keybind for butterfly wings on `b`
  * Warning should be printed
  * The existing keybind for `b` for other characters is not removed
* Add a non-character global for vomit3 on `v`
  * This should overwrite the keybind for everyone
* Press `b` in default mode
* Butterfly wings mutation is activated
* Press `b` twice
* Butterfly wings mutation is deactivated and then reactivated
* Press `v` in default mode
* vomit3 mutation is activated
* Save+load
* Keybinds persist
* Load different character, with no vomit3 mutation
* `b` key is bound to "go southwest"
* Press `v`
* "Character doesn't have this mutation" message is printed
* Load different character, in a world where vomit3 mutation is blacklisted (and another world, where this mutation is never loaded)
* Press `v`
* Nothing happens (no crash or debug messages, but regular messages are fine)

#### Additional context
